### ISSUE TITLE
Surprising behavior with Regex module with `'$'` 

### ISSUE THREAD

**Issue by pierce314159** [chapel-lang/chapel#20431]
### Summary of Problem

It seems `'$'` is not binding to the end of the string. In the example below, I expected a zero length match at the end of the string, i.e.
```
matches('$')
((matched = true, byteOffset = 10, numBytes = 0),)

search('$')
(matched = true, byteOffset = 10, numBytes = 0)
```
But instead it returns a zero length match at every position as if I had used `''` as my pattern
```
matches('$')
((matched = true, byteOffset = 0, numBytes = 0),) ((matched = true, byteOffset = 1, numBytes = 0),) ((matched = true, byteOffset = 2, numBytes = 0),) ((matched = true, byteOffset = 3, numBytes = 0),) ((matched = true, byteOffset = 4, numBytes = 0),) ((matched = true, byteOffset = 5, numBytes = 0),) ((matched = true, byteOffset = 6, numBytes = 0),) ((matched = true, byteOffset = 7, numBytes = 0),) ((matched = true, byteOffset = 8, numBytes = 0),) ((matched = true, byteOffset = 9, numBytes = 0),) ((matched = true, byteOffset = 10, numBytes = 0),)

search('$')
(matched = true, byteOffset = 0, numBytes = 0)
```

### Steps to Reproduce

**Source Code:**

```chapel
module PlayingAround {
  use Regex;
  use IO.FormattedIO;

  proc main () throws {
    var s = 'i string i';
    var pattern = '$';
    var myRegex = compile(pattern);
    writeln("matches('%s')".format(pattern));
    writeln(myRegex.matches(s));
    writeln();
    writeln("search('%s')".format(pattern));
    writeln(myRegex.search(s));
  }
}
```

**Compile command:**

```
chpl PlayingAround.chpl -o playing
```

**Execution command:**

```
 ./playing
```

### Configuration Information

- Output of `chpl --version`:
chpl version 1.27.0
  built with LLVM version 14.0.0
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled *
CHPL_LLVM: bundled *
CHPL_AUX_FILESYS: none
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0

**Comment by lydia-duncan** [comment]
`matches` is an iterator that will yield every match it finds.  It is based on RE2's `Match`, [documented here](https://github.com/google/re2/blob/main/re2/re2.h#L558).  In [our documentation](https://chapel-lang.org/docs/latest/modules/standard/Regex.html#Regex.regex.matches), it says that it takes an argument specifying the maximum number of matches to find, but I would assume that won't get what you're looking for either, since it seems like you want the last match in the string?

**Comment by pierce314159** [comment]
> `matches` is an iterator that will yield every match it finds. It is based on RE2's `Match`, [documented here](https://github.com/google/re2/blob/main/re2/re2.h#L558). In [our documentation](https://chapel-lang.org/docs/latest/modules/standard/Regex.html#Regex.regex.matches), it says that it takes an argument specifying the maximum number of matches to find, but I would assume that won't get what you're looking for either, since it seems like you want the last match in the string?

yeah I'm more saying that this isn't the behavior that i would expect from `matches` or `search`. I would think that both of these functions should only match on the last position. Based on what I see with `pattern='$'`, I would expect the output of something like `pattern = '.$'` to return multiple matches, but output is
```
matches('.$')
((matched = true, byteOffset = 9, numBytes = 1),)

search('.$')
(matched = true, byteOffset = 9, numBytes = 1)
```

I might be incorrect in this thinking, but I think of `search` as finding the first possible match and `matches` as iterating through every offset of string and applying the `match` function. The result im seeing the issue varies from what i would expect based on this
```python
>>> import re
>>> re.match('$', 'i string i')
>>> re.match('$', ' string i')
>>> re.match('$', 'string i')
>>> re.match('$', 'tring i')
# ...
>>> re.match('$', 'i')
>>> re.match('$', '')
<re.Match object; span=(0, 0), match=''>

>>> re.search('$', 'i string i')
<re.Match object; span=(10, 10), match=''>
```

**Comment by lydia-duncan** [comment]
I don't have a good answer for you, I think we're beholden to RE2's implementation here

**Comment by bradcray** [comment]
@lydia-duncan: Are you confident that the behavior we're generating today matches RE2's and that it isn't our own wrapping of it that's introducing a bug?

**Comment by lydia-duncan** [comment]
I think so.  I took a look at the implementation for matches and didn't see us doing anything that would cause us to alter the order of what was being returned or where it was looking or anything.

**Comment by lydia-duncan** [comment]
But I'd definitely feel more confident if someone else double checked

**Comment by bradcray** [comment]
I only looked enough to feel nervous about the combination of this while loop:

```chapel
    while nFound < maxMatches && cur <= endPos {
```

and this expression:

```chapel
      cur = matches[0].offset + max(1, matches[0].len);
```

In particular, it didn't make me confident that we were only returning the same matches that RE2 did/would.

**Comment by bradcray** [comment]
Hi Lydia —

I got more confidence in the module's implementation of `matches()`, but there's also a runtime shim between that module code and the re2 implementation in `runtime/src/qio/regex/bundled/re2-interface.cc`.  Did you look through that runtime code as part of your check of the implementation?

**Comment by bradcray** [comment]
@vasslitvinov — I know you've been looking at regex issues recently...  Is this one on your radar at all?  My memory thought 'yes', but I'm not seeing any links to work you're doing here, so maybe I'm wrong...

**Comment by vasslitvinov** [comment]
I did not see this... putting it in my sprint backlog.

Related: in #20864

**Comment by pierce314159** [comment]
Thanks so much for fixing this!!

**Comment by bradcray** [comment]
Ditto, @jabraham17!  (and thanks for reporting it to us, @pierce314159!)
